custom_function_double_backward.py

```diff
@@ -0,0 +1,206 @@
+# in the backward computation.
+#
+# Intermediate results should not be directly saved and used in backward though.
+# Because forward is performed in no-grad mode, if an intermediate result
+# of the forward pass is used to compute gradients in the backward pass
+# the backward graph of the gradients would not include the operations
+# that computed the intermediate result. This leads to incorrect gradients.
+class Sinh2(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        expx = torch.exp(x)
+        expnegx = torch.exp(-x)
+        ctx.save_for_backward(expx, expnegx)
+        # In order to be able to save the intermediate results, a trick is to
+        # include them as our outputs, so that the backward graph is constructed
+        return (expx - expnegx) / 2, expx, expnegx
+
+    @staticmethod
+    def backward(ctx, grad_out, _grad_out_exp, _grad_out_negexp):
+        expx, expnegx = ctx.saved_tensors
+        grad_input = grad_out * (expx + expnegx) / 2
+        # We cannot skip accumulating these even though we won't use the outputs
+        # directly. They will be used later in the second backward.
+        grad_input += _grad_out_exp * expx
+        grad_input -= _grad_out_exp * expnegx
+        return grad_input
+
+def sinh2(x):
+    # Create a wrapper that only returns the first output
+    return Sinh2.apply(x)[0]
+
+x = torch.rand(3, 3, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(sinh2, x)
+torch.autograd.gradgradcheck(sinh2, x)
+
+out = sinh2(x)
+grad_x, = torch.autograd.grad(out.sum(), x, create_graph=True)
+torchviz.make_dot((grad_x, x, out), params={"grad_x": grad_x, "x": x, "out": out})
+
+######################################################################
+# Finally, let's consider an example when it may not be possible for
+# autograd to track gradients for a functions backward at all.
+# We can imagine cube_backward to be a function that may require a
+# non-PyTorch library like SciPy or NumPy, or written as a
+# C++ extension. The workaround demonstrated here is to create another
+# custom function CubeBackward where you also manually specify the
+# backward of cube_backward!
+
+def cube_forward(x):
+    return x**3
+
+def cube_backward(grad_out, x):
+    return grad_out * 3 * x**2
+
+def cube_backward_backward(grad_out, sav_grad_out, x):
+    return grad_out * sav_grad_out * 6 * x
+
+def cube_backward_backward_grad_out(grad_out, x):
+    return grad_out * 3 * x**2
+
+class Cube(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        ctx.save_for_backward(x)
+        return cube_forward(x)
+
+    @staticmethod
+    def backward(ctx, grad_out):
+        x, = ctx.saved_tensors
+        return CubeBackward.apply(grad_out, x)
+
+class CubeBackward(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, grad_out, x):
+        ctx.save_for_backward(x, grad_out)
+        return cube_backward(grad_out, x)
+
+    @staticmethod
```

Comment:
maybe make this one once differentiable so that triple gradient will properly fail under that assumption that the backward backward functions are not differentiable either.

Reply:
good point!